src/Mod/Part/App/TopoShapeExpansion.cpp

```diff
@@ -2362,4 +2398,230 @@ bool TopoShape::fixSolidOrientation()
+{
+    return makeElementBoolean(maker, std::vector<TopoShape>(1, shape), op, tolerance);
+}
+
+
+TopoShape& TopoShape::makeElementBoolean(const char* maker,
+                                         const std::vector<TopoShape>& shapes,
+                                         const char* op,
+                                         double tolerance)
+{
+#if OCC_VERSION_HEX <= 0x060800
+    if (tolerance > 0.0) {
+        Standard_Failure::Raise("Fuzzy Booleans are not supported in this version of OCCT");
+    }
+#endif
+
+    if (!maker) {
+        FC_THROWM(Base::CADKernelError, "no maker");
+    }
+
+    if (!op) {
+        op = maker;
+    }
+
+    if (shapes.empty()) {
+        FC_THROWM(NullShapeException, "Null shape");
+    }
+
+    if (strcmp(maker, Part::OpCodes::Compound) == 0) {
+        return makeElementCompound(shapes, op, SingleShapeCompoundCreationPolicy::returnShape);
+    }
+    else if (boost::starts_with(maker, Part::OpCodes::Face)) {
+        std::string prefix(Part::OpCodes::Face);
+        prefix += '.';
+        const char* face_maker = 0;
+        if (boost::starts_with(maker, prefix)) {
+            face_maker = maker + prefix.size();
+        }
+        return makeElementFace(shapes, op, face_maker);
+    }
+    else if (strcmp(maker, Part::OpCodes::Wire) == 0) {
+        return makeElementWires(shapes, op);
+    }
+    else if (strcmp(maker, Part::OpCodes::Compsolid) == 0) {
+        BRep_Builder builder;
+        TopoDS_CompSolid Comp;
+        builder.MakeCompSolid(Comp);
+        for (auto& s : shapes) {
+            if (!s.isNull()) {
+                builder.Add(Comp, s.getShape());
+            }
+        }
+        setShape(Comp);
+        mapSubElement(shapes, op);
+        return *this;
+    }
+
+    if (strcmp(maker, Part::OpCodes::Pipe) == 0) {
+        if (shapes.size() != 2) {
+            FC_THROWM(Base::CADKernelError, "Not enough input shapes");
+        }
+        if (shapes[0].isNull() || shapes[1].isNull()) {
+            FC_THROWM(Base::CADKernelError, "Cannot sweep along empty spine");
+        }
+        if (shapes[0].getShape().ShapeType() != TopAbs_WIRE) {
+            FC_THROWM(Base::CADKernelError, "Spine shape is not a wire");
+        }
+        BRepOffsetAPI_MakePipe mkPipe(TopoDS::Wire(shapes[0].getShape()), shapes[1].getShape());
+        return makeElementShape(mkPipe, shapes, op);
+    }
+
+    if (strcmp(maker, Part::OpCodes::Shell) == 0) {
+        BRep_Builder builder;
+        TopoDS_Shell shell;
+        builder.MakeShell(shell);
+        for (auto& s : shapes) {
+            builder.Add(shell, s.getShape());
+        }
+        setShape(shell);
+        mapSubElement(shapes, op);
+        BRepCheck_Analyzer check(shell);
+        if (!check.IsValid()) {
+            ShapeUpgrade_ShellSewing sewShell;
+            setShape(sewShell.ApplySewing(shell), false);
+            // TODO confirm the above won't change OCCT topological naming
```

Comment:
```suggestion
            // TODO: confirm the above won't change OCCT topological naming
```
(so it gets picked up by automated tools)